File: src/logger.ts
import { appendFile, rename, stat, unlink } from "node:fs/promises";

const MAX_LOG_BYTES = 2 * 1024 * 1024;
const ROTATIONS = 3;

export function sanitizeText(value: string, maxLength = 1_000): string {
  return value
    .replace(/(authorization\s*[:=]\s*bearer\s+)[^\s,;]+/gi, "$1[REDACTED]")
    .replace(/((?:api[-_ ]?key|token)\s*[:=]\s*)[^\s,;]+/gi, "$1[REDACTED]")
    .slice(0, maxLength);
}

export class Logger {
  private writeChain = Promise.resolve();

  constructor(private readonly path: string) {}

  async initialize(): Promise<void> {
    try {
      const info = await stat(this.path);
      if (info.size < MAX_LOG_BYTES) {
        return;
      }
    } catch {
      return;
    }

    await unlink(`${this.path}.${ROTATIONS}`).catch(() => undefined);
    for (let index = ROTATIONS - 1; index >= 1; index -= 1) {
      await rename(`${this.path}.${index}`, `${this.path}.${index + 1}`).catch(() => undefined);
    }
    await rename(this.path, `${this.path}.1`).catch(() => undefined);
  }

  log(message: string): void {
    const line = `${new Date().toISOString()} ${sanitizeText(message)}\n`;
    this.writeChain = this.writeChain
      .then(() => appendFile(this.path, line, "utf8"))
      .catch(() => undefined);
  }
}

File: logger.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"sync"
	"time"
)

const maxLogBytes int64 = 2 * 1024 * 1024
const logRotations = 3

var bearerSecret = regexp.MustCompile(`(?i)(authorization\s*[:=]\s*bearer\s+)[^\s,;]+`)
var namedSecret = regexp.MustCompile(`(?i)((api[-_ ]?key|token)\s*[:=]\s*)[^\s,;]+`)

func sanitizeText(value string) string {
	value = bearerSecret.ReplaceAllString(value, "${1}[REDACTED]")
	value = namedSecret.ReplaceAllString(value, "${1}[REDACTED]")
	runes := []rune(value)
	if len(runes) > 1000 {
		value = string(runes[:1000])
	}
	return value
}

type fileLogger struct {
	path string
	mu   sync.Mutex
}

func newLogger(path string) *fileLogger { return &fileLogger{path: path} }

func (l *fileLogger) Initialize() error {
	info, err := os.Stat(l.path)
	if err != nil || info.Size() < maxLogBytes {
		return nil
	}
	_ = os.Remove(fmt.Sprintf("%s.%d", l.path, logRotations))
	for index := logRotations - 1; index >= 1; index-- {
		_ = os.Rename(fmt.Sprintf("%s.%d", l.path, index), fmt.Sprintf("%s.%d", l.path, index+1))
	}
	return os.Rename(l.path, l.path+".1")
}

func (l *fileLogger) Log(message string) {
	l.mu.Lock()
	defer l.mu.Unlock()
	file, err := os.OpenFile(l.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return
	}
	defer file.Close()
	line := time.Now().UTC().Format(time.RFC3339Nano) + " " + sanitizeText(strings.TrimRight(message, "\r\n")) + "\n"
	_, _ = file.WriteString(line)
}
